.github/api_best_practices.md

```diff
@@ -0,0 +1,231 @@
+e.g. `ReportConfigService`, `RbacService`. 
+All gRPC methods grouped into a single service **must** generally pertain to the primary resource of the service.
+
+### gRPC Method Name
+Methods **should** be named such that they provide insights into the functionality.
+
+Let us look at a few examples.`StartComplianceScan`, `RunComplianceScan`, and `GetComplianceScan` are not the same. 
+
+- `StartComplianceScan` **should** return without waiting for the compliance scan to complete. 
+- `RunComplianceScan` is ambiguous because it is unclear if the call waits for the scan to complete. 
+The ambiguity can be removed by adding a field to the request that helps clarify the expectation 
+e.g. `bool wait_for_scan_completion` if set to `true` informs the method to wait for the compliance 
+scan to complete. However, for long-running processes, it is **recommended** to create a job that 
+finishes the process asynchronously and return the job ID to the users which can be tracked via 
+dedicated job tracking method.
+- `GetComplianceScan` **should** not run a compliance scan but only fetches a stored one.
+
+Typically, the method name **should** follow the _VerbNoun_ convention.
+
+| Verb   | Noun           | Method name         |
+|--------|----------------|---------------------|
+| List   | Deployment     | `ListDeployments`   |
+| Get    | Deployment     | `GetDeployment`     |
+| Update | Deployment     | `UpdateDeployment`  |
+| Delete | Deployment     | `DeleteDeployment`  |
+| Notify | Violation      | `NotifyViolation`   |
+| Run    | ComplianceScan | `RunComplianceScan` |
+
+It is **recommended** that the verbs be imperative instead of inquisitive. Generally, the noun **should** be the resource type. 
+In some cases, the noun portion could be composed of multiple nouns e.g. `GetVulnerabilityDeferralState`, `RunPolicyScan`.
+
+| Inquisitive               | Imperative                      |
+|---------------------------|---------------------------------|
+| `IsRunComplete`           | `GetRunStatus`                  |
+| `IsAdmin`                 | `GetUserRole`                   |
+| `IsVulnerabilityDeferred` | `GetVulnerabilityDeferralState` |
+
+The noun portion of methods that act on a single resource **must** be singular e.g. `GetDeployment`. Those methods that 
+act on the collection of resources **must** be plural e.g. `ListDeployments`, `DeleteDeployments`. Avoid prepositions 
+(e.g. for, by) in method names as much as possible. Typically, this can be addressed by using a distinct verb, 
+adding a field to the request message, or restructuring _VerbNoun_.
+
+<table>
+<tr>
+<td><b>Instead of</b></td><td><b>Use</b></td>
+</tr>
+<tr>
+<td>
+
+`GetBaselineGeneratedNetworkPolicyForDeployment`
+
+</td>
+<td>
+
+```
+GenerateDeploymentNetworkPolicy
+
+GenerateDeploymentNetworkPolicyRequest {
+  bool from_baseline;  
+  bool from_network_flows;
+}
```

Comment:
While I get what you try to illustrate here (don't litter the URL, use query or body instead), all aspects of examples in guidelines are important. I doubt we want to encourage use of bool fields, so I recommend (also note correct formatting and proper proto syntax):
```
message GenerateDeploymentNetworkPolicyRequest {
    enum Source {
        UNKNOWN       = 0;
        BASELINE      = 1;
        NETWORK_FLOWS = 2;
    }

    Source from = 1;
}
```

Reply:
I specifically did not choose the enum path because it has the limitation that it does not provide setting multiple options.  For example, a network policy generation can consider both baseline and current net pols. This is where graphQL has a clear advantage.